polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -789,362 +665,335 @@ impl FragmentTree {
-					}
-				},
-			};
+		if let Some(output_head_hash) = output_head_hash {
+			if self.by_output_head.contains_key(&output_head_hash) {
+				// this is not a chain, there are multiple paths to the same state.
+				return true;
+			}
 
-			depth += 1;
+			// trivial 0-length cycle.
+			if parent_head_hash == output_head_hash {
+				return true;
+			}
+
+			// this should catch any other cycles. our output state cannot already be the parent
+			// state of another candidate, unless this is a cycle, since the already added
+			// candidates form a chain.
+			if self.by_parent_head.contains_key(&output_head_hash) {
+				return true;
+			}
 		}
 
-		Some(last_node)
+		false
 	}
 
-	// Find a node from the given iterator which is present in the ancestors
-	// collection. If there are multiple such nodes, return an error and log a warning. We don't
-	// accept forks in a parachain to be backed. The supplied ancestors should all form a chain.
-	// If there is no such node, return None.
-	fn find_valid_child<'a>(
+	// Checks the potential of a candidate to be added to the chain in the future.
+	// Verifies that the relay parent is in scope and not moving backwards and that we're not
+	// introducing forks or cycles with other candidates in the chain.
+	// `output_head_hash` is optional because we sometimes make this check before retrieving the
+	// collation.
+	fn check_potential(
 		&self,
-		ancestors: &'a mut Ancestors,
-		nodes: impl Iterator<Item = &'a (NodePointer, CandidateHash)> + 'a,
-	) -> Result<Option<NodePointer>, ()> {
-		let mut possible_children =
-			nodes.filter_map(|(node_ptr, hash)| match ancestors.remove(&hash) {
-				true => Some(node_ptr),
-				false => None,
-			});
-
-		// We don't accept forks in a parachain to be backed. The supplied ancestors
-		// should all form a chain.
-		let next = possible_children.next();
-		if let Some(second_child) = possible_children.next() {
-			if let (Some(NodePointer::Storage(first_child)), NodePointer::Storage(second_child)) =
-				(next, second_child)
-			{
-				gum::error!(
-					target: LOG_TARGET,
-					para_id = ?self.scope.para,
-					relay_parent = ?self.scope.relay_parent,
-					"Trying to find new backable candidates for a parachain for which we've backed a fork.\
-					This is a bug and the runtime should not have allowed it.\n\
-					Backed candidates with the same parent: {}, {}",
-					self.nodes[*first_child].candidate_hash,
-					self.nodes[*second_child].candidate_hash,
-				);
-			}
+		relay_parent: &Hash,
+		parent_head_hash: Hash,
+		output_head_hash: Option<Hash>,
+	) -> bool {
+		if self.is_fork_or_cycle(parent_head_hash, output_head_hash) {
+			return false;
+		}
 
-			Err(())
-		} else {
-			Ok(next.copied())
+		let Some(earliest_rp) = self.earliest_relay_parent() else { return false };
+
+		let Some(relay_parent) = self.scope.ancestor(relay_parent) else { return false };
+
+		if relay_parent.number < earliest_rp.number {
+			return false // relay parent moved backwards.
```

Comment:
Does it make sense to log trace here just in case for debugging ?

Reply:
I don't think it helps, because this function is also called when pruning the candidate storage on a new leaf update.
therefore, it's bound to happen most of the times under normal circumstances